src/client/Client.cc

```diff
@@ -1108,8 +1108,15 @@ void Client::insert_readdir_results(MetaRequest *request, MetaSession *session,
     string readdir_start = dirp->last_name;
 
     unsigned last_hash = 0;
-    if (!readdir_start.empty())
-      last_hash = ceph_frag_value(diri->hash_dentry_name(readdir_start));
+    if (hash_order) {
+      if (!readdir_start.empty()) {
+	last_hash = ceph_frag_value(diri->hash_dentry_name(readdir_start));
+      } else if (flags & CEPH_READDIR_OFFSET_HASH) {
+	/* mds understands offset_hash */
+	assert(readdir_offset == 2);
```

Comment:
Am I right in thinking that this assertion is saying "if readdir_start is empty then readdir_offset should be 2"?

If so (and it has nothing to do with the new OFFSET_HASH flag), then perhaps move the assertion up to after we assign readdir start, and write it as ``assert(!readdir_start.empty() || readdir_offset == 2)`` so that it's clearer to read